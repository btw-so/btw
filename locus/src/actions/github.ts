import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from 'modules/helpers';

import { ActionTypes } from 'literals';

export interface GetReposSuccessMeta {
  cached: boolean;
  query: string;
  updatedAt: number;
}

export const getRepos = createAction<string>(ActionTypes.GITHUB_GET_REPOS_REQUEST);
export const getReposSuccess = createAction(
  ActionTypes.GITHUB_GET_REPOS_SUCCESS,
  (payload: Record<string, any>[], meta: GetReposSuccessMeta) => actionPayload(payload, meta),
);
export const getReposFailure = createAction(
  ActionTypes.GITHUB_GET_REPOS_FAILURE,
  (payload: string, query: string) => actionPayload(payload, { query }),
);
