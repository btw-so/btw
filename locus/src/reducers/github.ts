import { createReducer } from '@reduxjs/toolkit';

import { topic } from 'config';
import { STATUS } from 'literals';

import { getRepos, getReposFailure, getReposSuccess } from 'actions';

import { GitHubState } from 'types';

export const githubState: GitHubState = {
  topics: {},
};

export default {
  github: createReducer<GitHubState>(githubState, builder => {
    builder
      .addCase(getRepos, (draft, { payload }) => {
        draft.topics[payload] = draft.topics[payload] || { ...topic };

        draft.topics[payload].message = '';
        draft.topics[payload].status = STATUS.RUNNING;
      })
      .addCase(getReposSuccess, (draft, { meta, payload }) => {
        const { cached = false, query = '', updatedAt = 0 } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].cached = cached;
        draft.topics[query].data = payload;
        draft.topics[query].status = STATUS.SUCCESS;
        draft.topics[query].updatedAt = updatedAt;
      })
      .addCase(getReposFailure, (draft, { meta, payload }) => {
        const { query = '' } = meta || {};

        draft.topics[query] = draft.topics[query] || { ...topic };

        draft.topics[query].message = payload;
        draft.topics[query].status = STATUS.ERROR;
        draft.topics[query].updatedAt = 0;
      });
  }),
};
