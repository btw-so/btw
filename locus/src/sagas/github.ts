import { now, request } from '@gilbarbara/helpers';
import { all, call, put, select, takeLatest } from 'redux-saga/effects';

import { hasValidCache } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { getRepos, getReposFailure, getReposSuccess } from 'actions';

/**
 * Get Repos
 *
 * @param {Object} action
 *
 */
export function* getReposSaga({ payload }: ReturnType<typeof getRepos>) {
  try {
    const { data = [], updatedAt = 0 } = yield select(s => s.github.topics?.[payload] || {});
    const hasCache = hasValidCache(updatedAt);
    let items;

    if (!hasCache) {
      ({ items = [] } = yield call(
        request,
        `https://api.github.com/search/repositories?q=${payload}&sort=stars`,
      ));
    }

    yield put(
      getReposSuccess(items || data, { cached: hasCache, query: payload, updatedAt: now() }),
    );
  } catch (error: any) {
    yield put(getReposFailure(error.message, payload));
  }
}

/**
 * GitHub Sagas
 */
export default function* root() {
  yield all([takeLatest(ActionTypes.GITHUB_GET_REPOS_REQUEST, getReposSaga)]);
}
