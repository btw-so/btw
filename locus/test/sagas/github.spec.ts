import { expectSaga } from 'redux-saga-test-plan';

import { ActionTypes } from 'literals';

import githubReducer from 'reducers/github';
import github, { getReposSaga } from 'sagas/github';

import githubRepos from 'test/__fixtures__/github-repos.json';
import { mergeState } from 'test-utils';

describe('github', () => {
  it('should have the expected watchers', () =>
    expectSaga(github)
      .run({ silenceTimeout: true })
      .then(result => {
        expect(result.toJSON()).toMatchSnapshot();
      }));

  describe('getRepos', () => {
    const initialAction = {
      type: ActionTypes.GITHUB_GET_REPOS_REQUEST,
      payload: 'react',
    };
    const initialState = () => ({
      github: githubReducer.github(undefined, initialAction),
    });

    it('should handle SUCCESS', () => {
      fetchMock.mockResponse(JSON.stringify({ items: githubRepos.items.slice(0, 2) }));

      return expectSaga(getReposSaga, {
        type: ActionTypes.GITHUB_GET_REPOS_REQUEST,
        payload: 'react',
      })
        .withReducer(initialState)
        .run()
        .then(result => {
          expect(result.toJSON()).toMatchSnapshot();
        });
    });

    it('should handle SUCCESS with cache', () => {
      fetchMock.mockResponse(JSON.stringify({ items: githubRepos.items.slice(0, 2) }));

      return expectSaga(getReposSaga, {
        type: ActionTypes.GITHUB_GET_REPOS_REQUEST,
        payload: 'react',
      })
        .withReducer(
          mergeState({
            github: {
              topics: {
                react: {
                  ...initialState().github.topics.react,
                  data: githubRepos.items.slice(0, 2),
                  updatedAt: 12345680000,
                },
              },
            },
          }),
        )
        .run()
        .then(result => {
          expect(result.toJSON()).toMatchSnapshot();
        });
    });

    it('should handle FAILURE', () => {
      fetchMock.mockReject(new Error('Failed to fetch'));

      return expectSaga(getReposSaga, {
        type: ActionTypes.GITHUB_GET_REPOS_REQUEST,
        payload: 'react',
      })
        .withReducer(initialState)
        .run()
        .then(result => {
          expect(result.toJSON()).toMatchSnapshot();
        });
    });
  });
});
