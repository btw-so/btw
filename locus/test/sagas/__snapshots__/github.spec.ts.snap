// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`github getRepos should handle FAILURE 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://api.github.com/search/repositories?q=react&sort=stars",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/request",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "meta": Object {
            "query": "react",
          },
          "payload": "Failed to fetch",
          "type": "GITHUB_GET_REPOS_FAILURE",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`github getRepos should handle SUCCESS 1`] = `
Object {
  "call": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "https://api.github.com/search/repositories?q=react&sort=stars",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/request",
      },
      "type": "CALL",
    },
  ],
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "meta": Object {
            "cached": false,
            "query": "react",
            "updatedAt": 1234567890,
          },
          "payload": Array [
            Object {
              "archive_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/{archive_format}{/ref}",
              "archived": false,
              "assignees_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/assignees{/user}",
              "blobs_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/blobs{/sha}",
              "branches_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/branches{/branch}",
              "clone_url": "https://github.com/freeCodeCamp/freeCodeCamp.git",
              "collaborators_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/collaborators{/collaborator}",
              "comments_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/comments{/number}",
              "commits_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/commits{/sha}",
              "compare_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/compare/{base}...{head}",
              "contents_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/contents/{+path}",
              "contributors_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/contributors",
              "created_at": "2014-12-24T17:49:19Z",
              "default_branch": "master",
              "deployments_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/deployments",
              "description": "freeCodeCamp.org's open source codebase and curriculum. Learn to code at home.",
              "disabled": false,
              "downloads_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/downloads",
              "events_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/events",
              "fork": false,
              "forks": 25018,
              "forks_count": 25018,
              "forks_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/forks",
              "full_name": "freeCodeCamp/freeCodeCamp",
              "git_commits_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/commits{/sha}",
              "git_refs_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/refs{/sha}",
              "git_tags_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/tags{/sha}",
              "git_url": "git://github.com/freeCodeCamp/freeCodeCamp.git",
              "has_downloads": true,
              "has_issues": true,
              "has_pages": true,
              "has_projects": false,
              "has_wiki": false,
              "homepage": "https://contribute.freecodecamp.org",
              "hooks_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/hooks",
              "html_url": "https://github.com/freeCodeCamp/freeCodeCamp",
              "id": 28457823,
              "issue_comment_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/issues/comments{/number}",
              "issue_events_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/issues/events{/number}",
              "issues_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/issues{/number}",
              "keys_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/keys{/key_id}",
              "labels_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/labels{/name}",
              "language": "JavaScript",
              "languages_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/languages",
              "license": Object {
                "key": "bsd-3-clause",
                "name": "BSD 3-Clause \\"New\\" or \\"Revised\\" License",
                "node_id": "MDc6TGljZW5zZTU=",
                "spdx_id": "BSD-3-Clause",
                "url": "https://api.github.com/licenses/bsd-3-clause",
              },
              "merges_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/merges",
              "milestones_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/milestones{/number}",
              "mirror_url": null,
              "name": "freeCodeCamp",
              "node_id": "MDEwOlJlcG9zaXRvcnkyODQ1NzgyMw==",
              "notifications_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/notifications{?since,all,participating}",
              "open_issues": 295,
              "open_issues_count": 295,
              "owner": Object {
                "avatar_url": "https://avatars0.githubusercontent.com/u/9892522?v=4",
                "events_url": "https://api.github.com/users/freeCodeCamp/events{/privacy}",
                "followers_url": "https://api.github.com/users/freeCodeCamp/followers",
                "following_url": "https://api.github.com/users/freeCodeCamp/following{/other_user}",
                "gists_url": "https://api.github.com/users/freeCodeCamp/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/freeCodeCamp",
                "id": 9892522,
                "login": "freeCodeCamp",
                "node_id": "MDEyOk9yZ2FuaXphdGlvbjk4OTI1MjI=",
                "organizations_url": "https://api.github.com/users/freeCodeCamp/orgs",
                "received_events_url": "https://api.github.com/users/freeCodeCamp/received_events",
                "repos_url": "https://api.github.com/users/freeCodeCamp/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/freeCodeCamp/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/freeCodeCamp/subscriptions",
                "type": "Organization",
                "url": "https://api.github.com/users/freeCodeCamp",
              },
              "private": false,
              "pulls_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/pulls{/number}",
              "pushed_at": "2020-11-08T00:02:53Z",
              "releases_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/releases{/id}",
              "score": 1,
              "size": 151476,
              "ssh_url": "git@github.com:freeCodeCamp/freeCodeCamp.git",
              "stargazers_count": 316482,
              "stargazers_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/stargazers",
              "statuses_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/statuses/{sha}",
              "subscribers_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/subscribers",
              "subscription_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/subscription",
              "svn_url": "https://github.com/freeCodeCamp/freeCodeCamp",
              "tags_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/tags",
              "teams_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/teams",
              "trees_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/trees{/sha}",
              "updated_at": "2020-11-07T23:51:21Z",
              "url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp",
              "watchers": 316482,
              "watchers_count": 316482,
            },
            Object {
              "archive_url": "https://api.github.com/repos/facebook/react/{archive_format}{/ref}",
              "archived": false,
              "assignees_url": "https://api.github.com/repos/facebook/react/assignees{/user}",
              "blobs_url": "https://api.github.com/repos/facebook/react/git/blobs{/sha}",
              "branches_url": "https://api.github.com/repos/facebook/react/branches{/branch}",
              "clone_url": "https://github.com/facebook/react.git",
              "collaborators_url": "https://api.github.com/repos/facebook/react/collaborators{/collaborator}",
              "comments_url": "https://api.github.com/repos/facebook/react/comments{/number}",
              "commits_url": "https://api.github.com/repos/facebook/react/commits{/sha}",
              "compare_url": "https://api.github.com/repos/facebook/react/compare/{base}...{head}",
              "contents_url": "https://api.github.com/repos/facebook/react/contents/{+path}",
              "contributors_url": "https://api.github.com/repos/facebook/react/contributors",
              "created_at": "2013-05-24T16:15:54Z",
              "default_branch": "master",
              "deployments_url": "https://api.github.com/repos/facebook/react/deployments",
              "description": "A declarative, efficient, and flexible JavaScript library for building user interfaces.",
              "disabled": false,
              "downloads_url": "https://api.github.com/repos/facebook/react/downloads",
              "events_url": "https://api.github.com/repos/facebook/react/events",
              "fork": false,
              "forks": 31471,
              "forks_count": 31471,
              "forks_url": "https://api.github.com/repos/facebook/react/forks",
              "full_name": "facebook/react",
              "git_commits_url": "https://api.github.com/repos/facebook/react/git/commits{/sha}",
              "git_refs_url": "https://api.github.com/repos/facebook/react/git/refs{/sha}",
              "git_tags_url": "https://api.github.com/repos/facebook/react/git/tags{/sha}",
              "git_url": "git://github.com/facebook/react.git",
              "has_downloads": true,
              "has_issues": true,
              "has_pages": true,
              "has_projects": true,
              "has_wiki": true,
              "homepage": "https://reactjs.org",
              "hooks_url": "https://api.github.com/repos/facebook/react/hooks",
              "html_url": "https://github.com/facebook/react",
              "id": 10270250,
              "issue_comment_url": "https://api.github.com/repos/facebook/react/issues/comments{/number}",
              "issue_events_url": "https://api.github.com/repos/facebook/react/issues/events{/number}",
              "issues_url": "https://api.github.com/repos/facebook/react/issues{/number}",
              "keys_url": "https://api.github.com/repos/facebook/react/keys{/key_id}",
              "labels_url": "https://api.github.com/repos/facebook/react/labels{/name}",
              "language": "JavaScript",
              "languages_url": "https://api.github.com/repos/facebook/react/languages",
              "license": Object {
                "key": "mit",
                "name": "MIT License",
                "node_id": "MDc6TGljZW5zZTEz",
                "spdx_id": "MIT",
                "url": "https://api.github.com/licenses/mit",
              },
              "merges_url": "https://api.github.com/repos/facebook/react/merges",
              "milestones_url": "https://api.github.com/repos/facebook/react/milestones{/number}",
              "mirror_url": null,
              "name": "react",
              "node_id": "MDEwOlJlcG9zaXRvcnkxMDI3MDI1MA==",
              "notifications_url": "https://api.github.com/repos/facebook/react/notifications{?since,all,participating}",
              "open_issues": 621,
              "open_issues_count": 621,
              "owner": Object {
                "avatar_url": "https://avatars3.githubusercontent.com/u/69631?v=4",
                "events_url": "https://api.github.com/users/facebook/events{/privacy}",
                "followers_url": "https://api.github.com/users/facebook/followers",
                "following_url": "https://api.github.com/users/facebook/following{/other_user}",
                "gists_url": "https://api.github.com/users/facebook/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/facebook",
                "id": 69631,
                "login": "facebook",
                "node_id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "organizations_url": "https://api.github.com/users/facebook/orgs",
                "received_events_url": "https://api.github.com/users/facebook/received_events",
                "repos_url": "https://api.github.com/users/facebook/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/facebook/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/facebook/subscriptions",
                "type": "Organization",
                "url": "https://api.github.com/users/facebook",
              },
              "private": false,
              "pulls_url": "https://api.github.com/repos/facebook/react/pulls{/number}",
              "pushed_at": "2020-11-07T23:58:11Z",
              "releases_url": "https://api.github.com/repos/facebook/react/releases{/id}",
              "score": 1,
              "size": 160472,
              "ssh_url": "git@github.com:facebook/react.git",
              "stargazers_count": 158686,
              "stargazers_url": "https://api.github.com/repos/facebook/react/stargazers",
              "statuses_url": "https://api.github.com/repos/facebook/react/statuses/{sha}",
              "subscribers_url": "https://api.github.com/repos/facebook/react/subscribers",
              "subscription_url": "https://api.github.com/repos/facebook/react/subscription",
              "svn_url": "https://github.com/facebook/react",
              "tags_url": "https://api.github.com/repos/facebook/react/tags",
              "teams_url": "https://api.github.com/repos/facebook/react/teams",
              "trees_url": "https://api.github.com/repos/facebook/react/git/trees{/sha}",
              "updated_at": "2020-11-07T23:22:09Z",
              "url": "https://api.github.com/repos/facebook/react",
              "watchers": 158686,
              "watchers_count": 158686,
            },
          ],
          "type": "GITHUB_GET_REPOS_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`github getRepos should handle SUCCESS with cache 1`] = `
Object {
  "put": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "action": Object {
          "meta": Object {
            "cached": true,
            "query": "react",
            "updatedAt": 1234567890,
          },
          "payload": Array [
            Object {
              "archive_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/{archive_format}{/ref}",
              "archived": false,
              "assignees_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/assignees{/user}",
              "blobs_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/blobs{/sha}",
              "branches_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/branches{/branch}",
              "clone_url": "https://github.com/freeCodeCamp/freeCodeCamp.git",
              "collaborators_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/collaborators{/collaborator}",
              "comments_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/comments{/number}",
              "commits_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/commits{/sha}",
              "compare_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/compare/{base}...{head}",
              "contents_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/contents/{+path}",
              "contributors_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/contributors",
              "created_at": "2014-12-24T17:49:19Z",
              "default_branch": "master",
              "deployments_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/deployments",
              "description": "freeCodeCamp.org's open source codebase and curriculum. Learn to code at home.",
              "disabled": false,
              "downloads_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/downloads",
              "events_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/events",
              "fork": false,
              "forks": 25018,
              "forks_count": 25018,
              "forks_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/forks",
              "full_name": "freeCodeCamp/freeCodeCamp",
              "git_commits_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/commits{/sha}",
              "git_refs_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/refs{/sha}",
              "git_tags_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/tags{/sha}",
              "git_url": "git://github.com/freeCodeCamp/freeCodeCamp.git",
              "has_downloads": true,
              "has_issues": true,
              "has_pages": true,
              "has_projects": false,
              "has_wiki": false,
              "homepage": "https://contribute.freecodecamp.org",
              "hooks_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/hooks",
              "html_url": "https://github.com/freeCodeCamp/freeCodeCamp",
              "id": 28457823,
              "issue_comment_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/issues/comments{/number}",
              "issue_events_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/issues/events{/number}",
              "issues_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/issues{/number}",
              "keys_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/keys{/key_id}",
              "labels_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/labels{/name}",
              "language": "JavaScript",
              "languages_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/languages",
              "license": Object {
                "key": "bsd-3-clause",
                "name": "BSD 3-Clause \\"New\\" or \\"Revised\\" License",
                "node_id": "MDc6TGljZW5zZTU=",
                "spdx_id": "BSD-3-Clause",
                "url": "https://api.github.com/licenses/bsd-3-clause",
              },
              "merges_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/merges",
              "milestones_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/milestones{/number}",
              "mirror_url": null,
              "name": "freeCodeCamp",
              "node_id": "MDEwOlJlcG9zaXRvcnkyODQ1NzgyMw==",
              "notifications_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/notifications{?since,all,participating}",
              "open_issues": 295,
              "open_issues_count": 295,
              "owner": Object {
                "avatar_url": "https://avatars0.githubusercontent.com/u/9892522?v=4",
                "events_url": "https://api.github.com/users/freeCodeCamp/events{/privacy}",
                "followers_url": "https://api.github.com/users/freeCodeCamp/followers",
                "following_url": "https://api.github.com/users/freeCodeCamp/following{/other_user}",
                "gists_url": "https://api.github.com/users/freeCodeCamp/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/freeCodeCamp",
                "id": 9892522,
                "login": "freeCodeCamp",
                "node_id": "MDEyOk9yZ2FuaXphdGlvbjk4OTI1MjI=",
                "organizations_url": "https://api.github.com/users/freeCodeCamp/orgs",
                "received_events_url": "https://api.github.com/users/freeCodeCamp/received_events",
                "repos_url": "https://api.github.com/users/freeCodeCamp/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/freeCodeCamp/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/freeCodeCamp/subscriptions",
                "type": "Organization",
                "url": "https://api.github.com/users/freeCodeCamp",
              },
              "private": false,
              "pulls_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/pulls{/number}",
              "pushed_at": "2020-11-08T00:02:53Z",
              "releases_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/releases{/id}",
              "score": 1,
              "size": 151476,
              "ssh_url": "git@github.com:freeCodeCamp/freeCodeCamp.git",
              "stargazers_count": 316482,
              "stargazers_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/stargazers",
              "statuses_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/statuses/{sha}",
              "subscribers_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/subscribers",
              "subscription_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/subscription",
              "svn_url": "https://github.com/freeCodeCamp/freeCodeCamp",
              "tags_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/tags",
              "teams_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/teams",
              "trees_url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp/git/trees{/sha}",
              "updated_at": "2020-11-07T23:51:21Z",
              "url": "https://api.github.com/repos/freeCodeCamp/freeCodeCamp",
              "watchers": 316482,
              "watchers_count": 316482,
            },
            Object {
              "archive_url": "https://api.github.com/repos/facebook/react/{archive_format}{/ref}",
              "archived": false,
              "assignees_url": "https://api.github.com/repos/facebook/react/assignees{/user}",
              "blobs_url": "https://api.github.com/repos/facebook/react/git/blobs{/sha}",
              "branches_url": "https://api.github.com/repos/facebook/react/branches{/branch}",
              "clone_url": "https://github.com/facebook/react.git",
              "collaborators_url": "https://api.github.com/repos/facebook/react/collaborators{/collaborator}",
              "comments_url": "https://api.github.com/repos/facebook/react/comments{/number}",
              "commits_url": "https://api.github.com/repos/facebook/react/commits{/sha}",
              "compare_url": "https://api.github.com/repos/facebook/react/compare/{base}...{head}",
              "contents_url": "https://api.github.com/repos/facebook/react/contents/{+path}",
              "contributors_url": "https://api.github.com/repos/facebook/react/contributors",
              "created_at": "2013-05-24T16:15:54Z",
              "default_branch": "master",
              "deployments_url": "https://api.github.com/repos/facebook/react/deployments",
              "description": "A declarative, efficient, and flexible JavaScript library for building user interfaces.",
              "disabled": false,
              "downloads_url": "https://api.github.com/repos/facebook/react/downloads",
              "events_url": "https://api.github.com/repos/facebook/react/events",
              "fork": false,
              "forks": 31471,
              "forks_count": 31471,
              "forks_url": "https://api.github.com/repos/facebook/react/forks",
              "full_name": "facebook/react",
              "git_commits_url": "https://api.github.com/repos/facebook/react/git/commits{/sha}",
              "git_refs_url": "https://api.github.com/repos/facebook/react/git/refs{/sha}",
              "git_tags_url": "https://api.github.com/repos/facebook/react/git/tags{/sha}",
              "git_url": "git://github.com/facebook/react.git",
              "has_downloads": true,
              "has_issues": true,
              "has_pages": true,
              "has_projects": true,
              "has_wiki": true,
              "homepage": "https://reactjs.org",
              "hooks_url": "https://api.github.com/repos/facebook/react/hooks",
              "html_url": "https://github.com/facebook/react",
              "id": 10270250,
              "issue_comment_url": "https://api.github.com/repos/facebook/react/issues/comments{/number}",
              "issue_events_url": "https://api.github.com/repos/facebook/react/issues/events{/number}",
              "issues_url": "https://api.github.com/repos/facebook/react/issues{/number}",
              "keys_url": "https://api.github.com/repos/facebook/react/keys{/key_id}",
              "labels_url": "https://api.github.com/repos/facebook/react/labels{/name}",
              "language": "JavaScript",
              "languages_url": "https://api.github.com/repos/facebook/react/languages",
              "license": Object {
                "key": "mit",
                "name": "MIT License",
                "node_id": "MDc6TGljZW5zZTEz",
                "spdx_id": "MIT",
                "url": "https://api.github.com/licenses/mit",
              },
              "merges_url": "https://api.github.com/repos/facebook/react/merges",
              "milestones_url": "https://api.github.com/repos/facebook/react/milestones{/number}",
              "mirror_url": null,
              "name": "react",
              "node_id": "MDEwOlJlcG9zaXRvcnkxMDI3MDI1MA==",
              "notifications_url": "https://api.github.com/repos/facebook/react/notifications{?since,all,participating}",
              "open_issues": 621,
              "open_issues_count": 621,
              "owner": Object {
                "avatar_url": "https://avatars3.githubusercontent.com/u/69631?v=4",
                "events_url": "https://api.github.com/users/facebook/events{/privacy}",
                "followers_url": "https://api.github.com/users/facebook/followers",
                "following_url": "https://api.github.com/users/facebook/following{/other_user}",
                "gists_url": "https://api.github.com/users/facebook/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/facebook",
                "id": 69631,
                "login": "facebook",
                "node_id": "MDEyOk9yZ2FuaXphdGlvbjY5NjMx",
                "organizations_url": "https://api.github.com/users/facebook/orgs",
                "received_events_url": "https://api.github.com/users/facebook/received_events",
                "repos_url": "https://api.github.com/users/facebook/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/facebook/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/facebook/subscriptions",
                "type": "Organization",
                "url": "https://api.github.com/users/facebook",
              },
              "private": false,
              "pulls_url": "https://api.github.com/repos/facebook/react/pulls{/number}",
              "pushed_at": "2020-11-07T23:58:11Z",
              "releases_url": "https://api.github.com/repos/facebook/react/releases{/id}",
              "score": 1,
              "size": 160472,
              "ssh_url": "git@github.com:facebook/react.git",
              "stargazers_count": 158686,
              "stargazers_url": "https://api.github.com/repos/facebook/react/stargazers",
              "statuses_url": "https://api.github.com/repos/facebook/react/statuses/{sha}",
              "subscribers_url": "https://api.github.com/repos/facebook/react/subscribers",
              "subscription_url": "https://api.github.com/repos/facebook/react/subscription",
              "svn_url": "https://github.com/facebook/react",
              "tags_url": "https://api.github.com/repos/facebook/react/tags",
              "teams_url": "https://api.github.com/repos/facebook/react/teams",
              "trees_url": "https://api.github.com/repos/facebook/react/git/trees{/sha}",
              "updated_at": "2020-11-07T23:22:09Z",
              "url": "https://api.github.com/repos/facebook/react",
              "watchers": 158686,
              "watchers_count": 158686,
            },
          ],
          "type": "GITHUB_GET_REPOS_SUCCESS",
        },
        "channel": undefined,
      },
      "type": "PUT",
    },
  ],
  "select": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [],
        "selector": "@@redux-saga-test-plan/json/function/<anonymous>",
      },
      "type": "SELECT",
    },
  ],
}
`;

exports[`github should have the expected watchers 1`] = `
Object {
  "fork": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "args": Array [
          "GITHUB_GET_REPOS_REQUEST",
          "@@redux-saga-test-plan/json/function/getReposSaga",
        ],
        "context": null,
        "fn": "@@redux-saga-test-plan/json/function/takeLatest",
      },
      "type": "FORK",
    },
  ],
  "take": Array [
    Object {
      "@@redux-saga/IO": true,
      "combinator": false,
      "payload": Object {
        "pattern": "GITHUB_GET_REPOS_REQUEST",
      },
      "type": "TAKE",
    },
  ],
}
`;
